### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [2]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

In [3]:
purchase_data.head(2)

Purchase ID           SN  Age Gender  Item ID  \
0            0      Lisim78   20   Male      108   
1            1  Lisovynya38   40   Male      143   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56

In [4]:
purchase_data.describe(include="all")

Purchase ID         SN         Age Gender     Item ID     Item Name  \
count    780.000000        780  780.000000    780  780.000000           780   
unique          NaN        576         NaN      3         NaN           179   
top             NaN  Lisosia93         NaN   Male         NaN  Final Critic   
freq            NaN          5         NaN    652         NaN            13   
mean     389.500000        NaN   22.714103    NaN   91.755128           NaN   
std      225.310896        NaN    6.659444    NaN   52.697702           NaN   
min        0.000000        NaN    7.000000    NaN    0.000000           NaN   
25%      194.750000        NaN   20.000000    NaN   47.750000           NaN   
50%      389.500000        NaN   22.000000    NaN   92.000000           NaN   
75%      584.250000        NaN   25.000000    NaN  138.000000           NaN   
max      779.000000        NaN   45.000000    NaN  183.000000           NaN   

             Price  
count   780.000000  
unique         NaN  
top            NaN  
freq           NaN  
mean      3.050987  
std       1.169549  
min       1.000000  
25%       1.980000  
50%       3.150000  
75%       4.080000  
max       4.990000

* Display the total number of players


In [5]:
#Computes for Total Players
totalplayer = len(purchase_data["SN"].value_counts())
#totalplayer
tplayer_df = pd.DataFrame([{"Total Players":totalplayer}])
tplayer_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [6]:
#Get unique item
num_uniqueitems = len(purchase_data["Item Name"].value_counts())
#num_uniqueitems
ave_price = purchase_data["Price"].mean()
#ave_price  how to round this in 2 decimal
num_purchase = purchase_data["Purchase ID"].count()
#num_purchase
total_revenue = round((ave_price)*(num_purchase),2)
#total_revenue

In [7]:
purchasing_analysis_total =pd.DataFrame([{"Number of Unique Items":num_uniqueitems,
                        "Average Price":ave_price,
                        "Number of Purchases":num_purchase,
                        "Total Revenue":total_revenue}])
purchasing_analysis_total
purchasing_analysis_total.style.format({"Average Price":"${:,.2f}","Total Revenue":"${:,.2f}"})

In [3]:
## EXPECTED OUTPUT

Number of Unique Items Average Price Number of Purchases Total Revenue
0                     179         $3.05                 780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [8]:
Count_Male_Players = purchase_data.loc[purchase_data["Gender"]=="Male",:]
#Count_Male_Players
Unique_Count_Male_Players= Count_Male_Players["SN"].nunique()
Unique_Count_Male_Players
percent_male_players =round(((Unique_Count_Male_Players) / (totalplayer))*100,2)
percent_male_players

84.03

In [9]:
Count_Female_Players = purchase_data.loc[purchase_data["Gender"]=="Female",:]
#Count_Female_Players
Unique_Count_Female_Players= Count_Female_Players["SN"].nunique()
Unique_Count_Female_Players
percent_female_players = round((Unique_Count_Female_Players / totalplayer) * 100,2)
percent_female_players

14.06

In [10]:
Count_Other_Players = purchase_data.loc[purchase_data["Gender"]=="Other / Non-Disclosed",:]
#Count_Other_Players
Unique_Count_Other_Players= Count_Other_Players["SN"].nunique()
Unique_Count_Other_Players
percent_other_players = round(((Unique_Count_Other_Players / totalplayer) * 100),2)
percent_other_players

1.91


Gender Demographics OUTPUT

In [11]:
Gender_df = pd.DataFrame({
    "Total Counts":[Unique_Count_Male_Players,Unique_Count_Female_Players,Unique_Count_Other_Players],
    "Percentage":[percent_male_players,percent_female_players,percent_other_players]
},index=['Male','Female','Other / Non-Disclosed'])

Gender_df


Total Counts  Percentage
Male                            484       84.03
Female                           81       14.06
Other / Non-Disclosed            11        1.91


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [12]:
#count how many purchases each gender
purchase_counts=purchase_data["Gender"].value_counts()
purchase_counts

Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [13]:
#create purchasing analysis df(pgender_df)
grp_Pgender_df = purchase_data.groupby('Gender')

#print(grp_Pgender_df)
#To visualized the data
grp_Pgender_df.count().head()

Purchase ID   SN  Age  Item ID  Item Name  Price
Gender                                                                 
Female                         113  113  113      113        113    113
Male                           652  652  652      652        652    652
Other / Non-Disclosed           15   15   15       15         15     15

In [14]:
#Get the average Purchase Price

AveragePrice=round(grp_Pgender_df["Price"].mean(),2)
AveragePrice

Gender
Female                   3.20
Male                     3.02
Other / Non-Disclosed    3.35
Name: Price, dtype: float64

In [15]:
#Get the Total Purchase Value

TPurchaseVal=round(grp_Pgender_df["Price"].sum(),2)
TPurchaseVal

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [16]:
#get average total purchage per person, use the unique count per gender
AveTotal_PurchPer=round((TPurchaseVal)/(Gender_df["Total Counts"]),2)
AveTotal_PurchPer

Female                   4.47
Male                     4.07
Other / Non-Disclosed    4.56
dtype: float64


Purchasing Analysis (Gender) OUTPUT

In [18]:
#Creating a Summary DataFrame  Purchasing Analysis
gendersummary_df=pd.DataFrame({"Purchase Count":purchase_counts,
                    "Average Purchase Price":AveragePrice.map("${:.2f}".format),
                     "Total Purchase Value":TPurchaseVal.map("${:.2f}".format),
                     "Avg Total Purchase per Person":AveTotal_PurchPer.map("${:.2f}".format)})
gendersummary_df

Purchase Count Average Purchase Price  \
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [19]:
print(purchase_data["Age"].max())
print(purchase_data["Age"].min())

45
7


In [20]:
#create bins 
bins = [0,9.9,14.9,19.9,24.9,29.9,34.9,39.9,45]
#create labels
age_labels=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

In [21]:
#slice the data and place it into bins
pd.cut(purchase_data["Age"], bins, labels=age_labels).head()

0    20-24
1      40+
2    20-24
3    20-24
4    20-24
Name: Age, dtype: category
Categories (8, object): ['<10' < '10-14' < '15-19' < '20-24' < '25-29' < '30-34' < '35-39' < '40+']

In [22]:
#place the data series into new column inside the Dataframe
purchase_data["Age Group"]=pd.cut(purchase_data["Age"], bins, labels=age_labels,include_lowest=True)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Group  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [23]:
#create a grpby object

agegrp= purchase_data.groupby('Age Group')['SN'].nunique()
agegrp

Age Group
<10       17
10-14     22
15-19    107
20-24    258
25-29     77
30-34     52
35-39     31
40+       12
Name: SN, dtype: int64

In [24]:
#Compute for the Age Grps Percentage 
#totalplayer is the count of unique player

agegrp_percent = purchase_data.groupby('Age Group')['SN'].nunique()/totalplayer
agegrp_percent_df=pd.DataFrame(agegrp_percent)
agegrp_percent_df

SN
Age Group          
<10        0.029514
10-14      0.038194
15-19      0.185764
20-24      0.447917
25-29      0.133681
30-34      0.090278
35-39      0.053819
40+        0.020833


Age Demographics OUTPUT

In [25]:
# Merge Data and Create Summary 

merge_agegrp_df=pd.merge(agegrp,agegrp_percent_df,on="Age Group")
merge_agegrp_df.columns=["Total Count","Percentage of Players"]
merge_agegrp_df["Percentage of Players"]=merge_agegrp_df["Percentage of Players"].map("{:.2%}".format)
merge_agegrp_df

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

In [6]:
## EXPECTED OUTPUT

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [27]:
#Copy existing binning for purchasing analysis
purchase_df = purchase_data
purchase_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Group  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [28]:
#Create groupby df and display purchase count

grp_purchase = purchase_df.groupby("Age Group")
grp_purchase_df = pd.DataFrame(grp_purchase["Purchase ID"].count())
grp_purchase_df

Purchase ID
Age Group             
<10                 23
10-14               28
15-19              136
20-24              365
25-29              101
30-34               73
35-39               41
40+                 13

In [29]:
#Computes for Average Purchase Price

ave_purchase_price = grp_purchase["Price"].mean()
grp_ave_purprice_df =pd.DataFrame(ave_purchase_price)
grp_ave_purprice_df

Price
Age Group          
<10        3.353478
10-14      2.956429
15-19      3.035956
20-24      3.052219
25-29      2.900990
30-34      2.931507
35-39      3.601707
40+        2.941538

In [30]:
#Computes for Total Purchase Value

total_purchase_val = grp_purchase["Price"].sum()
grp_total_purval_df =pd.DataFrame(total_purchase_val)
grp_total_purval_df

Price
Age Group         
<10          77.13
10-14        82.78
15-19       412.89
20-24      1114.06
25-29       293.00
30-34       214.00
35-39       147.67
40+          38.24

In [31]:
#Computes for Average Total Purchase per Person

avg_tot_purchperson = total_purchase_val/purchase_data.groupby('Age Group')['SN'].nunique()
grpavg_tot_purchperson_df = pd.DataFrame(avg_tot_purchperson)
grpavg_tot_purchperson_df

0
Age Group          
<10        4.537059
10-14      3.762727
15-19      3.858785
20-24      4.318062
25-29      3.805195
30-34      4.115385
35-39      4.763548
40+        3.186667


MERGING OF DF's AND SUMMARY OUTPUT

In [32]:
#Merge Purchage Count and Purchase Price 

merge_grp_avepurchase_df=pd.merge(grp_purchase_df,grp_ave_purprice_df,on="Age Group")
merge_grp_avepurchase_df.columns=["Purchase Count","Average Purchase Price"]
merge_grp_avepurchase_df["Average Purchase Price"]=merge_grp_avepurchase_df["Average Purchase Price"].astype("float").round(2)

merge_grp_avepurchase_df

Purchase Count  Average Purchase Price
Age Group                                        
<10                    23                    3.35
10-14                  28                    2.96
15-19                 136                    3.04
20-24                 365                    3.05
25-29                 101                    2.90
30-34                  73                    2.93
35-39                  41                    3.60
40+                    13                    2.94

In [33]:
#Merge Total Purchase Value to the previous DF

merge_total_purval_df=pd.merge(merge_grp_avepurchase_df,grp_total_purval_df,on="Age Group")
merge_total_purval_df.columns=["Purchase Count","Average Purchase Price","Total Purchase Value"]
merge_total_purval_df

Purchase Count  Average Purchase Price  Total Purchase Value
Age Group                                                              
<10                    23                    3.35                 77.13
10-14                  28                    2.96                 82.78
15-19                 136                    3.04                412.89
20-24                 365                    3.05               1114.06
25-29                 101                    2.90                293.00
30-34                  73                    2.93                214.00
35-39                  41                    3.60                147.67
40+                    13                    2.94                 38.24

In [34]:
#Merge the Avg Total Purcahse per Person and Create a Summary

summary_purchasing_df = pd.merge(merge_total_purval_df,grpavg_tot_purchperson_df,on="Age Group")
summary_purchasing_df.columns=["Purchase Count","Average Purchase Price","Total Purchase Value","Avg Total Purchase per Person"]
summary_purchasing_df["Avg Total Purchase per Person"] = summary_purchasing_df["Avg Total Purchase per Person"].astype("float").round(2)
summary_purchasing_df
summary_purchasing_df.style.format({"Average Purchase Price":"${:,.2f}", "Total Purchase Value":"${:,.2f}","Avg Total Purchase per Person":"${:,.2f}"})

In [7]:
##EXPECTED OUTPUT

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                              
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [35]:
#Get the Purchase Count
grpSN_purchase = purchase_df.groupby("SN")
grpSN_purchase_df= pd.DataFrame(grpSN_purchase["Purchase ID"].count())
grpSN_purchase_df

Purchase ID
SN                        
Adairialis76             1
Adastirin33              1
Aeda94                   1
Aela59                   1
Aelaria33                1
...                    ...
Yathecal82               3
Yathedeu43               2
Yoishirrala98            1
Zhisrisu83               2
Zontibe81                3

[576 rows x 1 columns]

In [36]:
#Get the Average Price

SN_ave_purchase_price = grpSN_purchase["Price"].mean()
grp_SN_ave_purprice_df =pd.DataFrame(SN_ave_purchase_price).astype("float").round(2)
grp_SN_ave_purprice_df

Price
SN                  
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
...              ...
Yathecal82      2.07
Yathedeu43      3.01
Yoishirrala98   4.58
Zhisrisu83      3.94
Zontibe81       2.68

[576 rows x 1 columns]

In [37]:
#Compute for Total Purchase Value

SN_total_purchase_value=grpSN_purchase["Price"].sum()
grp_SN_total_purval_df=pd.DataFrame(SN_total_purchase_value)
grp_SN_total_purval_df

Price
SN                  
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
...              ...
Yathecal82      6.22
Yathedeu43      6.02
Yoishirrala98   4.58
Zhisrisu83      7.89
Zontibe81       8.03

[576 rows x 1 columns]

MERGING DF & SUMMARY DF

In [38]:
merge_SNcount_price_df=pd.merge(grpSN_purchase_df,grp_SN_ave_purprice_df,on="SN")
merge_SNcount_price_df.columns=["Purchase Count","Average Purchase Price"]
merge_SNcount_price_df

Purchase Count  Average Purchase Price
SN                                                   
Adairialis76                1                    2.28
Adastirin33                 1                    4.48
Aeda94                      1                    4.91
Aela59                      1                    4.32
Aelaria33                   1                    1.79
...                       ...                     ...
Yathecal82                  3                    2.07
Yathedeu43                  2                    3.01
Yoishirrala98               1                    4.58
Zhisrisu83                  2                    3.94
Zontibe81                   3                    2.68

[576 rows x 2 columns]

In [39]:
merge_SNtotal_purval_df=pd.merge(merge_SNcount_price_df,grp_SN_total_purval_df,on="SN")
merge_SNtotal_purval_df.columns=["Purchase Count","Average Purchase Price","Total Purchase Value"]
merge_SNtotal_purval_df

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                         
Adairialis76                1                    2.28                  2.28
Adastirin33                 1                    4.48                  4.48
Aeda94                      1                    4.91                  4.91
Aela59                      1                    4.32                  4.32
Aelaria33                   1                    1.79                  1.79
...                       ...                     ...                   ...
Yathecal82                  3                    2.07                  6.22
Yathedeu43                  2                    3.01                  6.02
Yoishirrala98               1                    4.58                  4.58
Zhisrisu83                  2                    3.94                  7.89
Zontibe81                   3                    2.68                  8.03

[576 rows x 3 columns]


TOP SPENDER OUTPUT

In [41]:
sumary_topspender_df = merge_SNtotal_purval_df.sort_values("Total Purchase Value", ascending=False).head()
#sumary_topspender_df.head()
sumary_topspender_df.style.format({"Average Purchase Price":"${:,.2f}","Total Purchase Value":"${:,.2f}"})

In [8]:
##EXPECTED OUTPUT

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [47]:
#Retrieve columns from purchase_df
columns_popular = purchase_df[["Item ID","Item Name","Price"]]
#columns_popular

In [48]:
#Create Grp
grp_poplar_itemid = columns_popular.groupby(["Item ID", "Item Name"])
#grp_poplar_itemid.head()

In [49]:
purchase_count = grp_poplar_itemid["Price"].count()
tot_purval = grp_poplar_itemid["Price"].sum()
item_price = tot_purval / purchase_count
grp_popular_items = pd.DataFrame({"Purchase Count":purchase_count, 
                                  "Item Price" : item_price,
                                  "Total Purchase Value" : tot_purval
                                 })
grp_popular_items.head()

Purchase Count  Item Price  Total Purchase Value
Item ID Item Name                                                           
0       Splinter                         4      1.2800                  5.12
1       Crucifer                         4      2.9425                 11.77
2       Verdict                          6      2.4800                 14.88
3       Phantomlight                     6      2.4900                 14.94
4       Bloodlord's Fetish               5      1.7000                  8.50


MOST POPULAR ITEMS OUTPUT

In [50]:
#SORT Per Purchase Count 
grp_popular_items_sort = grp_popular_items.sort_values("Purchase Count", ascending=False).head()
#grp_popular_items_sort.head()
grp_popular_items_sort.style.format({"Item Price":"${:,.2f}", "Total Purchase Value":"${:,.2f}"})

In [59]:
EXPECTED OUTPUT

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [51]:
grp_profit_items = grp_popular_items
grp_profit_items.head()

Purchase Count  Item Price  Total Purchase Value
Item ID Item Name                                                           
0       Splinter                         4      1.2800                  5.12
1       Crucifer                         4      2.9425                 11.77
2       Verdict                          6      2.4800                 14.88
3       Phantomlight                     6      2.4900                 14.94
4       Bloodlord's Fetish               5      1.7000                  8.50


MOST PROFITABLE ITEMS OUTPUT

In [52]:
grp_profit_items_sort = grp_profit_items.sort_values("Total Purchase Value", ascending=False).head()
#grp_profit_items_sort.head()
grp_profit_items_sort.style.format({"Item Price":"${:,.2f}", "Total Purchase Value":"${:,.2f}"})

In [10]:
EXPECTED OUTPUT

Purchase Count  \
Item ID Item Name                                                     
92      Final Critic                                             13   
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
103     Singed Scalpel                                            8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

 Based on the Output Data, We can say that :
 
 Out of 576 players, 484 players(84.03%) are males with the Total Purchase value of $1967.64

  Player ages between 20-24 years old has the highest purchase count, total of 365 purchases with Total Purchase Value of $1,114.06

  Final Critic is the most popular and profitable item.
